### PR TITLE
ISSUE-1106 Add middleware examples

### PR DESCRIPTION
**Pull request by adammorr**
https://github.com/justeattakeaway/JustSaying/issues/1106

Add an example console app to show how middlewares can be used, as well as calling out some default behaviours / pitfalls.

There's a bit of code overlap with the other example so might be able to consolidate or, leave as two separate things!

Heavily inspired by what George put together 🎆

### CODE REVIEW

File: samples/src/JustSaying.Sample.Middleware/JustSaying.Sample.Middleware.csproj [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    <PackageReference Include="Microsoft.Extensions.Hosting" Version="7.0.1" />
+    <PackageReference Include="Microsoft.Extensions.Logging.Console" Version="7.0.0" />
+    <PackageReference Include="Polly" Version="7.2.3" />
+    <PackageReference Include="Serilog.AspNetCore" Version="6.1.0" />
+  </ItemGroup>
+  
+  <ItemGroup>
+    <None Update="appsettings.json">
+      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+    </None>
```
Comment (martincostello) [inline samples/src/JustSaying.Sample.Middleware/JustSaying.Sample.Middleware.csproj:27]:
```suggestion
    <None Update="appsettings.json" CopyToOutputDirectory="Always" />
```

File: samples/src/JustSaying.Sample.Middleware/Middlewares/PollyJustSayingMiddleware.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        _policy = CreateMessageRetryPolicy();
+    }
+
+    protected override async Task<bool> RunInnerAsync(HandleMessageContext context, Func<CancellationToken, Task<bool>> func, CancellationToken stoppingToken)
+    {
+        Log.Information("[{MiddlewareName}] Started {MessageType}", nameof(PollyJustSayingMiddleware), context.Message.GetType().Name);
+
+        try
+        {
+            return await _policy.ExecuteAsync(() => func(stoppingToken));
```
Comment (martincostello) [inline samples/src/JustSaying.Sample.Middleware/Middlewares/PollyJustSayingMiddleware.cs:26]: With Polly it's best practice to also allow Polly's internals access to the `CancellationToken`.

```suggestion
            return await _policy.ExecuteAsync(() => func(stoppingToken), stoppingToken);
```

File: samples/src/JustSaying.Sample.Middleware/Program.cs [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+using JustSaying.Sample.Middleware.Extensions;
+using JustSaying.Sample.Middleware.Handlers;
+using JustSaying.Sample.Middleware.Messages;
+using JustSaying.Sample.Middleware.Middlewares;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Serilog;
+
+Console.Title = "Middleware Example";
```
Comment (martincostello) [inline samples/src/JustSaying.Sample.Middleware/Program.cs:13]: Nit: this should be set back to the existing value in a `finally` when the application terminates.

File: samples/src/JustSaying.Sample.Middleware/README.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+The purpose of this application is to demonstrate the usage of middleware in JustSaying.
+
+## Usage and Considerations
+
+By default, all subscriptions will have a standard set of a middleware added which will, among other things;
+
+* Wire up the `IHandlerAsync<T>` for your message
+* Emit logs and metrics for message handling
+* Deletion of a message if your handler returns successful
```
Comment (martincostello) [inline samples/src/JustSaying.Sample.Middleware/README.md:11]:
```suggestion
* Delete a message if your handler returns successfully
```

File: samples/src/JustSaying.Sample.Middleware/README.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    cfg.WithMiddlewareConfiguration(middlewareBuilder =>
+    {
+        middlewareBuilder.Use(new FirstMiddleware());
+        middlewareBuilder.UseDefaults<SampleMessage>(typeof(SampleMessageHandler)); // Add default middleware pipeline
+        middlewareBuilder.Use(new LastMiddleware());
+    });
+});
+```
+
+If you don't call `UseDefaults<T>(...)` then messages won't pass through your handler, nor will they be deleted from their queues so its strongly recommended to use it.
```
Comment (martincostello) [inline samples/src/JustSaying.Sample.Middleware/README.md:28]:
```suggestion
If you don't call `UseDefaults<T>(...)` then messages won't pass through your handler, nor will they be deleted from their queues, so its strongly recommended to use it.
```

File: samples/src/JustSaying.Sample.Middleware/Middlewares/PollyJustSayingMiddleware.cs (hunk 2 above)
Comment (adammorr) [reply samples/src/JustSaying.Sample.Middleware/Middlewares/PollyJustSayingMiddleware.cs:26]: Hmm this doesn't compile and doesn't look like there's an overload for passing in a CT 🤔

File: samples/src/JustSaying.Sample.Middleware/Program.cs (hunk 3 above)
Comment (adammorr) [reply samples/src/JustSaying.Sample.Middleware/Program.cs:13]: It's probably not needed really so I'll drop it

File: samples/src/JustSaying.Sample.Middleware/Middlewares/PollyJustSayingMiddleware.cs (hunk 2 above)
Comment (adammorr) [reply samples/src/JustSaying.Sample.Middleware/Middlewares/PollyJustSayingMiddleware.cs]: It was there, I just looked badly

**Review by martincostello — APPROVED** [review]

**Pull request merged**